### PR TITLE
feat(g05): add OpenGalaxea G0.5 policy integration

### PR DESCRIPTION
**Pull request by pkooij**
## Processor-owned tokenization update (2026-08-11)

- The serializable G0.5 input pipeline now builds native samples and runs `encode_inference()` / `encode_train()` before policy execution.
- The native backend consumes the emitted sequence IDs, labels, token types, split index, pixels, and samples; policy action/text/training methods no longer invoke input tokenization.
- Updated checkpoints: https://huggingface.co/pepijn223/g05-super-chatton-subtask/commit/b8a1645a09a09bec763511034b119081431d14a4, https://huggingface.co/lerobot/g05_base/commit/87b9ffc905cf7a3355301dcd3c575fc36addbbf1, https://huggingface.co/lerobot/g05_libero/commit/afc7a130ded54f2bc4ef4a8675f5935c7e399c39, https://huggingface.co/lerobot/g05_robotwin20/commit/10865322d57979fa74c722d9849a11462239e935, and https://huggingface.co/lerobot/g05_so101/commit/98f42427322e854ae4ae90a521a832b2694eaa57
- Validation: 54 passed, 3 skipped; touched-file pre-commit hooks pass.

## Final #4183 contract (2026-08-11)

- G0.5 explicitly returns True from supports_text_generation() and directly implements public generate_text(processed_batch) through its native System-2 CoT/action stream.
- Its input pipeline starts with one RenderMessagesStep; the policy-specific BBox image-size step remains only where G0.5 needs it.
- The BBox/Subtask recipe lives in G05Config; recipe_path is only an explicit external override.
- Updated checkpoints: https://huggingface.co/pepijn223/g05-super-chatton-subtask/commit/750531f47066863609a27a488bfbbfc8142cb111, https://huggingface.co/lerobot/g05_base/commit/2b87340f906d38f65f52b6fc8e0811e0ecf286ae, and https://huggingface.co/lerobot/g05_so101/commit/636b5192e2a82f3aceb1f33aafd036cc73d93e10
- Focused final suite: 55 passed, 1 skipped; all touched-file pre-commit hooks passed.

## Summary

OpenGalaxea G0.5 is a Qwen3.5-2B vision-language-action model that generates embodied reasoning and robot actions in one stream. LeRobot exposes direct action generation as System 1 and the optional embodied chain-of-thought path as System 2; both use the same model and, for System 2, the action remains conditioned on the post-reasoning KV state.

This PR adds a native LeRobot implementation of the G0.5 proprioception path, temporal multi-camera vision, action expert, continuous flow head, discrete ActionCodec head, prompt serialization, training losses, and unified CoT/action decode. It does not import the OpenGalaxea Python package or require Hydra, OmegaConf, or `PYTHONPATH` setup. Transformers supplies the Qwen3.5 decoder, vision blocks, tokenizer, cache behavior, and tied language head.

This PR is stacked on and depends on #4183 (`codex/language-runtime`) for the shared interactive language runtime and telemetry.

> [!WARNING]
> G0.5 code and checkpoints use the non-commercial G0.5 Community License. The converted checkpoints remain private/access-controlled under the `lerobot` organization.

## Converted checkpoints

| Profile | LeRobot checkpoint | Contract | Action path | Predicted → executed |
| --- | --- | --- | --- | ---: |
| Base / R1 Lite | [`lerobot/g05_base`](https://huggingface.co/lerobot/g05_base) | 27D whole-body | Flow or ActionCodec, optional System 2 | 32 → 16 |
| LIBERO | [`lerobot/g05_libero`](https://huggingface.co/lerobot/g05_libero) | 20D grouped policy layout | Continuous flow / System 1 | 32 → 10 |
| RoboTwin | [`lerobot/g05_robotwin20`](https://huggingface.co/lerobot/g05_robotwin20) | 20D grouped policy layout | Continuous flow / System 1 | 32 → 8 |
| SO-101 | [`lerobot/g05_so101`](https://huggingface.co/lerobot/g05_so101) | 20D `so100` embodiment layout | Flow or ActionCodec, optional System 2 | 32 → 16 |

Each checkpoint packages the LeRobot config, tokenizer/processor assets, normalization statistics, exact prompt template, model weights, license/notice files, and the ActionCodec sidecar where applicable.

## Model inputs, outputs, and mappings

| Profile | Inputs | Mapping into the model | Model output and postprocessing |
| --- | --- | --- | --- |
| Base / R1 Lite | Task; head + left/right wrist cameras; six-frame history; 27D proprioception | R1 Lite/Pro joints → released 27D whole-body layout; z-score-tail normalization with q01/q99 grippers | 32 × 27 continuous flow or discrete ActionCodec actions; first 16 executed |
| LIBERO | Task; exterior + right-wrist cameras; right EEF 6D + gripper | Raw 7D state/action ↔ 20D grouped policy layout; stepwise q01/q99 | 32 × 20 internal chunk mapped back to the 7D controller; first 10 executed |
| RoboTwin | Task; high + left/right-wrist cameras; two arms with 6D pose + gripper | Raw 14D dual-arm state/action ↔ 20D grouped policy layout; stepwise q01/q99 | 32 × 20 internal chunk mapped back to the 14D controller; first 8 executed |
| SO-101 | Task; exterior + right wrist; optional left wrist; six arm joints | Raw 6D joints ↔ 20D `so100` policy layout; absent left wrist is zero-filled; stepwise q01/q99 | 32 × 20 internal chunk mapped back to 6D joint commands; first 16 executed |

`G05Policy.forward` trains assistant BBox/Subtask/ActionCodec tokens with cross entropy and continuous actions with flow matching. Joint checkpoints optimize `ce_loss + fm_loss` through six native optimizer groups.

## Author parity

Author and native LeRobot paths used the same converted base checkpoint, prompt, six-frame image tensors, proprioception, action mask, BF16 CUDA execution, and RNG seed on an RTX 5090.

<details>
<summary>Parity script</summary>

```bash
set -euo pipefail
ROOT=/dev/shm/g05-validation/g05-base-r1lite
PY=/home/croissant/lerobot/.venv/bin/python
AUTHOR_SRC=/dev/shm/g05-validation/GalaxeaVLA/src
LEROBOT_SRC=/dev/shm/g05-validation/lerobot-pr/src
export ROOT

PYTHONPATH="$AUTHOR_SRC:$LEROBOT_SRC" "$PY" <<'PY'
import json, os, torch
from safetensors.torch import load_file
from g05.models.g05.g05_policy_qwen35 import G05PolicyQwen35

root = os.environ["ROOT"]
cfg = json.load(open(f"{root}/config.json"))
mc = cfg["author_model_config"]
mc["pretrained_model_path"] = None
mc["hf_processor_path"] = f"{root}/hf_processor"
mc["AT_CONFIG"]["ckpt_dir"] = f"{root}/action_tokenizer.pt"
mc.update(predict_cot=False, discrete_action=False, continuous_action=True, return_continuous_action=True)
model = G05PolicyQwen35(**mc)
state = {k.removeprefix("backend."): v for k, v in load_file(f"{root}/model.safetensors").items()}
state["model.vlm.output_proj.weight"] = state["model.vlm.input_proj.weight"]
model.load_state_dict(state, strict=True)
model.eval().to("cuda", dtype=torch.bfloat16)
model.apply_fp32_params()
sample = {
    "template": cfg["prompt_template"], "command": "pick up the red cup",
    "embodiment": cfg["embodiment"], "prompt": "predict subtask",
    "proprio": torch.zeros(1, 27, device="cuda"),
    **{f"image{i}": (256, 256) for i in range(3)},
}
batch = {
    "samples": [sample],
    "pixel_values": {k: torch.zeros(1, 6, 3, 256, 256, device="cuda") for k in cfg["camera_order"]},
    "action_dim_is_pad": torch.zeros(1, 27, dtype=torch.bool, device="cuda"),
}
torch.manual_seed(1234)
with torch.inference_mode(), torch.autocast("cuda", dtype=torch.bfloat16):
    action = model.predict_action(batch)["action"].float().cpu()
torch.save(action, "/tmp/g05-author-forward.pt")
PY

PYTHONPATH="$LEROBOT_SRC" "$PY" <<'PY'
import os, torch
from lerobot.policies.g05.modeling_g05 import G05Policy

root = os.environ["ROOT"]
policy = G05Policy.from_pretrained(root, strict=True).eval().to("cuda")
policy.config.predict_cot = False
policy.config.discrete_action = False
policy.config.continuous_action = True
policy.config.action_head = "flow"
policy.backend.model_config.update(predict_cot=False, discrete_action=False, continuous_action=True)
sample = {
    "template": policy.config.prompt_template, "command": "pick up the red cup",
    "embodiment": policy.config.embodiment, "prompt": "predict subtask",
    "proprio": torch.zeros(1, 27, device="cuda"),
    **{f"image{i}": (256, 256) for i in range(3)},
}
batch = {
    "samples": [sample],
    "pixel_values": {
        k: torch.zeros(1, 6, 3, 256, 256, device="cuda") for k in policy.config.camera_order
    },
    "action_dim_is_pad": torch.zeros(1, 27, dtype=torch.bool, device="cuda"),
}
torch.manual_seed(1234)
with torch.inference_mode(), torch.autocast("cuda", dtype=torch.bfloat16):
    action = policy.backend.predict_action(batch)["action"].float().cpu()
diff = (action - torch.load("/tmp/g05-author-forward.pt", weights_only=True)).abs()
print(
    f"forward(): shape={tuple(action.shape)} max_abs={diff.max().item():.8g} "
    f"mean_abs={diff.mean().item():.8g} p99_abs={torch.quantile(diff, .99).item():.8g}"
)
PY
```

</details>

### Results

| Check | Result |
| --- | ---: |
| Output shape | `(1, 32, 27)` |
| Maximum absolute difference | `0.028770089` |
| Mean absolute difference | `0.001175447` |
| p99 absolute difference | `0.008231522` |

Additional checkpoint validation:

- strict load of all 945 native model tensors;
- strict load of all 208 ActionCodec tensors;
- exact prompt token IDs and token-type masks;
- exact ActionCodec code IDs, with decoder mean absolute difference around `3e-7`;
- finite joint training losses (`ce_loss=13.8141`, `fm_loss=1.32647`) and gradients for all model tensors.

## Checkpoint loading

The G0.5 loader now constructs the 11 GB policy on the meta device, assigns safetensors directly on the target device, rebuilds non-persistent rotary buffers, and loads the ActionCodec sidecar afterwards. On the RTX 5090 host, strict `g05-base` loading reduced peak host RSS from **25,909,096 KiB** to **3,892,836 KiB** and wall time from **15.72 s** to **7.27 s**, with unchanged **5.85 GiB** final CUDA allocation and zero remaining meta tensors.

## Fresh final-native benchmarks

The published native checkpoints were evaluated on benchmark head `eb9a4dae7` on an NVIDIA RTX 5090. These results predate the checkpoint-faithful GatedDeltaNet numerical override in `318e45da6`; the benchmark rerun is pending.

| Checkpoint | Environment and protocol | Final-native result |
| --- | --- | ---: |
| [`lerobot/g05_libero`](https://huggingface.co/lerobot/g05_libero) | LIBERO-10; all 10 tasks; 5 episodes per task; seed 0 | **50/50 (100%)** |
| [`lerobot/g05_robotwin20`](https://huggingface.co/lerobot/g05_robotwin20) | RoboTwin `beat_block_hammer`; `demo_clean`; unseen instructions; 50 counted rollouts | **46/50 (92.0%)** |

LIBERO reported `avg_sum_reward=1.0`, `avg_max_reward=1.0`, and `pc_success=100.0`. RoboTwin completed all 50 counted rollouts without a planner hang, RPC failure, or CUDA failure; four genuine task failures remain in the denominator.

## SO-101 zero-shot follow-up

- [ ] Run [`lerobot/g05_so101`](https://huggingface.co/lerobot/g05_so101) zero-shot on a physical SO-101 with the interactive language runtime from #4183.

```bash
lerobot-rollout \
  --language \
  --policy.path=lerobot/g05_so101 \
  --policy.device=cuda \
  --robot.type=so101_follower \
  --robot.port=/dev/ttyACM0 \
  --robot.id=my_follower \
  --robot.cameras='{exterior: {type: opencv, index_or_path: 0, width: 640, height: 480, fps: 30}, wrist_right: {type: opencv, index_or_path: 2, width: 640, height: 480, fps: 30}}' \
  --task="pick up the cube" \
  --mode=paused \
  --ctrl_hz=30 \
  --chunk_hz=1
```

At the prompt, type `/action <task>`. Leave `--direct_subtask` unset to retain native System 2 CoT and telemetry.

## Checklist

- [x] Native LeRobot runtime without the OpenGalaxea package, Hydra, OmegaConf, or runtime `PYTHONPATH` setup
- [x] Converted base, LIBERO, RoboTwin, and SO-101 checkpoints packaged
- [x] Strict native model and ActionCodec loading
- [x] Serializable preprocessors/postprocessors and mapping coverage
- [x] Native training forward/backward and optimizer-group validation
- [x] Author/native parity script and results included
- [x] Focused G0.5 policy and direct language-runtime tests — 55 passed, 1 skipped
- [x] Ruff, formatting, lockfile, Bandit, and mypy checks
- [ ] GitHub CI has run on this stacked PR (currently no checks are reported)
- [ ] Rerun LIBERO and RoboTwin after the GatedDeltaNet numerical override (prior final-native results are retained above)
- [ ] Physical SO-101 zero-shot run
- [ ] Community review linked

### CODE REVIEW

**Comment by pkooij** [comment]
Parity rerun on the RTX 5090 after `318e45da6` (same converted base checkpoint, prompt, six-frame tensors, proprioception, BF16 autocast, and seed as the author path).

<details>
<summary>Parity script</summary>

```bash
#!/usr/bin/env bash
set -euo pipefail

ROOT=/dev/shm/g05-validation/g05-base-r1lite
PY=/home/croissant/lerobot/.venv/bin/python
AUTHOR_SRC=/dev/shm/g05-validation/GalaxeaVLA/src
BASELINE_SRC=/dev/shm/g05-load-baseline-svrTS0/src
EXPERIMENT_SRC=/dev/shm/g05-meta-parity-PPEiuS/src
export ROOT

PYTHONPATH="$AUTHOR_SRC:$EXPERIMENT_SRC" "$PY" <<'PY'
import json
import os

import torch
from g05.models.g05.g05_policy_qwen35 import G05PolicyQwen35
from safetensors.torch import load_file

root = os.environ["ROOT"]
with open(f"{root}/config.json") as config_file:
    config = json.load(config_file)
model_config = config["author_model_config"]
model_config["pretrained_model_path"] = None
model_config["hf_processor_path"] = f"{root}/hf_processor"
model_config["AT_CONFIG"]["ckpt_dir"] = f"{root}/action_tokenizer.pt"
model_config.update(
    predict_cot=False,
    discrete_action=False,
    continuous_action=True,
    return_continuous_action=True,
)
model = G05PolicyQwen35(**model_config)
state = {
    key.removeprefix("backend."): value
    for key, value in load_file(f"{root}/model.safetensors").items()
}
state["model.vlm.output_proj.weight"] = state["model.vlm.input_proj.weight"]
model.load_state_dict(state, strict=True)
model.eval().to("cuda", dtype=torch.bfloat16)
model.apply_fp32_params()
sample = {
    "template": config["prompt_template"],
    "command": "pick up the red cup",
    "embodiment": config["embodiment"],
    "prompt": "predict subtask",
    "proprio": torch.zeros(1, 27, device="cuda"),
    **{f"image{index}": (256, 256) for index in range(3)},
}
batch = {
    "samples": [sample],
    "pixel_values": {
        key: torch.zeros(1, 6, 3, 256, 256, device="cuda")
        for key in config["camera_order"]
    },
    "action_dim_is_pad": torch.zeros(1, 27, dtype=torch.bool, device="cuda"),
}
torch.manual_seed(1234)
with torch.inference_mode(), torch.autocast("cuda", dtype=torch.bfloat16):
    action = model.predict_action(batch)["action"].float().cpu()
torch.save(action, "/tmp/g05-author-forward-meta-experiment.pt")
print(f"author forward(): shape={tuple(action.shape)}")
PY

run_native() {
    local source_root="$1"
    local label="$2"
    PYTHONPATH="$source_root" "$PY" - "$label" <<'PY'
import os
import sys

import torch
from lerobot.policies.g05.modeling_g05 import G05Policy

root = os.environ["ROOT"]
label = sys.argv[1]
policy = G05Policy.from_pretrained(root, strict=True, local_files_only=True).eval().to("cuda")
policy.config.predict_cot = False
policy.config.discrete_action = False
policy.config.continuous_action = True
policy.config.action_head = "flow"
policy.backend.model_config.update(
    predict_cot=False,
    discrete_action=False,
    continuous_action=True,
)
sample = {
    "template": policy.config.prompt_template,
    "command": "pick up the red cup",
    "embodiment": policy.config.embodiment,
    "prompt": "predict subtask",
    "proprio": torch.zeros(1, 27, device="cuda"),
    **{f"image{index}": (256, 256) for index in range(3)},
}
batch = {
    "samples": [sample],
    "pixel_values": {
        key: torch.zeros(1, 6, 3, 256, 256, device="cuda")
        for key in policy.config.camera_order
    },
    "action_dim_is_pad": torch.zeros(1, 27, dtype=torch.bool, device="cuda"),
}
torch.manual_seed(1234)
with torch.inference_mode(), torch.autocast("cuda", dtype=torch.bfloat16):
    action = policy.backend.predict_action(batch)["action"].float().cpu()
reference = torch.load("/tmp/g05-author-forward-meta-experiment.pt", weights_only=True)
diff = (action - reference).abs()
print(
    f"{label} forward(): shape={tuple(action.shape)} max_abs={diff.max().item():.8g} "
    f"mean_abs={diff.mean().item():.8g} p99_abs={torch.quantile(diff, .99).item():.8g}"
)
PY
}

run_native "$BASELINE_SRC" baseline
run_native "$EXPERIMENT_SRC" experiment
```

</details>

```text
baseline forward(): shape=(1, 32, 27) max_abs=0.085852146 mean_abs=0.001576264 p99_abs=0.015251891
experiment forward(): shape=(1, 32, 27) max_abs=0.028770089 mean_abs=0.001175447 p99_abs=0.0082315216
```

Strict-load measurement on the same host: peak RSS `25,909,096 → 3,892,836 KiB`, wall time `15.72 → 7.27 s`, final CUDA allocation unchanged at `5.85 GiB`; all 18 linear-attention layers use the G0.5 override and no meta tensors remain.

**Comment by Maximellerbach** [comment]
Superseded by: #4445
